### PR TITLE
Python 3

### LINKED ISSUE #80: Add the support for python 3

### PR DESCRIPTION
**Pull request by stephanerobert**
Python 3 support for Fake Switches code.

Resolves #80 But as of this commit, the tftpy library does not officially support python 3

### CODE REVIEW

File: fake_switches/terminal/telnet.py [hunk 1]
```diff
@@ -60,14 +62,14 @@ def applicationDataReceived(self, data):
...
-        self.transport.write(lf_to_crlf(data))
+        self.transport.write(lf_to_crlf(data).encode())
 
     def writeln(self, data):
         self.write(data)
         self.next_line()
 
     def next_line(self):
-        self.write(CR + LF)
+        self.write('\n')
```
Comment (mat128) [inline fake_switches/terminal/telnet.py]: This changed. Was it intentional?

File: fake_switches/terminal/telnet.py [hunk 2]
```diff
@@ -130,9 +132,11 @@ def command(self, line):
             self.transport.loseConnection()
 
     def applicationDataReceived(self, data):
-        if data in self._printable_chars:
+        assert(isinstance(data, bytes))
```
Comment (mat128) [inline fake_switches/terminal/telnet.py]: you want it to crash unless it receives bytes or was that a test?

File: requirements.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
+-e git+https://github.com/stephanerobert/tftpy.git@master#egg=tftpy
 netaddr>=0.7.13
-tftpy>=0.6.1
 pycrypto>=2.6.1
-Twisted==15.5.0
+Twisted>=16.6.0, <17.0
```
Comment (mat128) [inline requirements.txt]: Can you rebase on Felix's latest work? It should get rid of that line.

File: tests/cisco/test_cisco_auto_enabled_switch.py [hunk 4]
```diff
@@ -30,18 +30,19 @@ def test_enable_command_requires_a_password(self, t):
         t.read("my_switch#")
 
     def create_client(self):
+        return TelnetTester("telnet", cisco_switch_ip, cisco_auto_enabled_switch_telnet_port, u'root', u'root')
```
Comment (mat128) [inline tests/cisco/test_cisco_auto_enabled_switch.py]: This seems like an error.

File: tests/cmd/test_main.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 class FakeSwitchesTest(unittest.TestCase):
     def test_fake_switches_entrypoint_cisco_generic(self):
         p = subprocess.Popen(get_base_args() + ['cisco_generic', TEST_BIND_HOST, TEST_BIND_PORT])
-
+        time.sleep(1)
```
Comment (mat128) [inline tests/cmd/test_main.py]: Was that intentional?

File: tests/util/global_reactor.py [hunk 6]
```diff
@@ -67,83 +67,83 @@ def start_reactor(cls):
         SwitchTelnetService(cisco_switch_ip, telnet_port=cisco_switch_telnet_port, switch_core=switch_core,
                             users={'root': 'root'}).hook_to_reactor(cls._threaded_reactor.reactor)
         SwitchSshService(cisco_switch_ip, ssh_port=cisco_switch_ssh_port, switch_core=switch_core,
-                         users={'root': 'root'}).hook_to_reactor(cls._threaded_reactor.reactor)
+                         users={'root': b'root'}).hook_to_reactor(cls._threaded_reactor.reactor)
```
Comment (mat128) [inline tests/util/global_reactor.py]: Username doesnt need it?

File: tests/util/protocol_util.py [hunk 7]
```diff
@@ -112,6 +113,6 @@ def get_ssh_connect_command(self):
     def login(self):
         self.wait_for("Username: ")
         self.write(self.username)
-        self.read("Password: ")
+        self.wait_for("[pP]assword: ", True)
```
Comment (mat128) [inline tests/util/protocol_util.py]: This is different (even though it should be as valid, I think).

**Comment by mat128** [comment]
Didnt we need any change in main.py concerning the username/password to pass to the core? It seems the core is now using unicode strings where I think the main is still passing basic strings.

File: fake_switches/terminal/telnet.py (hunk 1 above)
Comment (JoProvost) [inline fake_switches/terminal/telnet.py]: As you can see, lf_to_crlf() should handle this case so I think it is a good thing to use '\n' here.

File: fake_switches/terminal/telnet.py (hunk 2 above)
Comment (JoProvost) [inline fake_switches/terminal/telnet.py]: Effectively, it **has** to be bytes or `data.decode()` will fail.

File: tests/util/global_reactor.py (hunk 6 above)
Comment (stephanerobert) [inline tests/util/global_reactor.py]: No, I had to dig inside Twisted to understand what was happening

File: tests/cmd/test_main.py (hunk 5 above)
Comment (stephanerobert) [inline tests/cmd/test_main.py]: Yes. I was getting connection error without it.

File: tests/cisco/test_cisco_auto_enabled_switch.py (hunk 4 above)
Comment (stephanerobert) [inline tests/cisco/test_cisco_auto_enabled_switch.py]: Haha, It is. Testing leftovers ;-)

File: requirements.txt (hunk 3 above)
Comment (stephanerobert) [inline requirements.txt]: Yes. Will do

File: fake_switches/terminal/telnet.py (hunk 2 above)
Comment (stephanerobert) [inline fake_switches/terminal/telnet.py]: Good catch. I though I had removed them all.

File: fake_switches/terminal/telnet.py (hunk 1 above)
Comment (stephanerobert) [inline fake_switches/terminal/telnet.py]: Yes

**Pull request merged**